src/platform_linux/shmem.c

```diff
@@ -0,0 +1,1152 @@
+      if (shm->shm_num_frags_inuse > shm->shm_num_frags_inuse_HWM) {
+         shm->shm_num_frags_inuse_HWM = shm->shm_num_frags_inuse;
+      }
+
+      // We should really assert that the other fields are zero, but for now
+      // re-init this fragment tracker.
+      memset(frag, 0, sizeof(*frag));
+      frag->shm_frag_addr = addr;
+      frag->shm_frag_size = size;
+
+      frag->shm_frag_allocated_to_pid = getpid();
+      frag->shm_frag_allocated_to_tid = platform_get_tid();
+
+      // The freed_by_pid/freed_by_tid == 0 means fragment is still allocated.
+
+      // Track highest address of large-fragment that is being tracked.
+      if (shm->shm_large_frag_hip < addr) {
+         shm->shm_large_frag_hip = addr;
+      }
+      break;
+   }
+   shm_unlock_mem_frags(shm);
+}
+
+/*
+ * -----------------------------------------------------------------------------
+ * platform_shm_track_free() - See if this memory fragment being freed is
+ * already being tracked. If so, it's a large fragment allocation, which can be
+ * re-cycled after this free. Do the book-keeping accordingly to record that
+ * this large-fragment is no longer in-use and can be recycled.
+ * -----------------------------------------------------------------------------
+ */
+static void
+platform_shm_track_free(shmem_info *shm,
+                        void       *addr,
+                        const char *objname,
+                        const char *func,
+                        const char *file,
+                        const int   lineno)
+{
+   shm_lock_mem_frags(shm);
+   shm->shm_nfrees++;
+
+   // If we are freeing a fragment beyond the high-address of all
+   // large fragments tracked, then this is certainly not a large
+   // fragment. So, no further need to see if it's a tracked fragment.
+   if (addr > shm->shm_large_frag_hip) {
+      shm->shm_nf_search_skipped++; // Track # of optimizations done
+      shm_unlock_mem_frags(shm);
+      return;
+   }
+   bool found_tracked_frag = FALSE;
+   bool trace_shmem        = (Trace_shmem || Trace_shmem_frees);
+
+   shm_frag_info *frag = shm->shm_mem_frags;
+   for (int fctr = 0; fctr < ARRAY_SIZE(shm->shm_mem_frags); fctr++, frag++) {
+      if (!frag->shm_frag_addr || (frag->shm_frag_addr != addr)) {
+         continue;
+      }
+      found_tracked_frag = TRUE;
+
+      // Cross-check the recording we did when fragment was allocated
+      // We cannot check the tid as the parent process' tid may be 0.
+      // We could have come across a free fragment that was previously
+      // used by the parent process.
+      // debug_assert(frag->shm_frag_allocated_to_tid != 0);
+      debug_assert(frag->shm_frag_allocated_to_pid != 0);
+      debug_assert(frag->shm_frag_size != 0);
+
+      // Mark the fragment as in-use by recording the process/thread that's
+      // doing the free.
+      frag->shm_frag_freed_by_pid = getpid();
+      frag->shm_frag_freed_by_tid = platform_get_tid();
```

Comment:
All this `debug_assert` and freed_by_pid / allocated_to_pid business has worked very well.

With this all in-place from the get-go, I didn't run into a single bad code / erroneous large-free-fragment ownership in any of the high-volume benchmarking.